docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,557 @@
+
+A better example of what a very good peer would be is:
+A peer that has a list of timestamps whose length is equal to 10, and the oldest timestamp was a week ago and the newest from an hour ago. This will make the peer fall into "< 1 week ago" bucket but also all the other recent buckets.
+
+We suggest to also turn the "amount of times seen" into buckets, such that we sort peers into buckets of "how many times" they were seen, modifying the selection criteria for good peers to become:
+
+* Peers that fall in most recent buckets + Peers that fall in highest seen count buckets
+
+**The interface**: We will go with the name `Bucketer` for the lack of a better one for the moment
+
+```go
+type Bucketer interface {
+    AddPeer(peer.ID, time.Time)
+    AddPeers([]peer.ID, time.Time)
+    Sort()
+    BestPeers() []peer.ID
+    PurgePeer(peer.ID)
+    PurgeAll()
+}
+```
+
+The implementer of this interface should rely on golang maps as the native abstraction for buckets, example:
+
+```go
+type bucketer struct {
+    timeBuckets map[BucketDuration][]peer.ID
+    seenBuckets map[int][]peer.ID
+    peersSeen map[peer.ID][]time.Time
+}
+```
+
+with `BucketDuration` being a new type to enumerate all possible time buckets, example:
+
+```go
+type BucketDuration string
+const (
+    LTOneHour = "< 1 hour"
+    LTOneDay = "< 1 day"
+    ...
+)
+```
+
+The `AddPeer` or `AddPeers`, and the `Sort` methods are intended to be called from within a procedure `fetchSortPersistPeers` that is called within a routine that periodically fetches peers from the peer store of the host (_`peerstore.Peerstore`_) which contains all the peers we are connected to.
+
+`fetchSortPersistPeers` will also be called on node shutdown.
+
+`fetchSortPersistPeers` will be as follows:
+
+```go
+func fetchSortPersistPeers(bcktr Bucketer, store PeerStore, host host.Host) {
+    goodPeers := make([]peer.PeerInfo, 0)
+    peers := host.Peerstore().PeersWithAddrs()
+    bcktr.AddPeers(peers)
+    bcktr.Sort()
+    bestPeers = bcktr.BestPeers()
+    for _, bestpeer := range bestPeers {
+        goodPeers = append(goodPeers, host.Peerstore().PeerInfo(bestpeer))
+    }
+    store.Put(goodPeers)
+}
+```
+
+### Node Startup
+
+In node startup, we propose to use the same mechanism that is proposed in approach A, but also add changes to register the `fetchSortPersistPeers` routine to trigger periodically every other `filterBootstrappersInterval`.
+
+```diff
+    baseComponents = fx.Options(
+        ...
+        fx.Provide(newModule),
++   fx.Invoke(func (ctx context.Context, lc fx.Lifecycle, cfg Config, bcktr Bucketer, pstore PeerStore, host host.Host) {
++     ch := make(chan struct{})
++     ticker := time.NewTicker(cfg.filterBootstrappersInterval * time.Millisecond)
++     go func() {
++       select {
++         case <-ctx.Done():
++         case <-ch:
++         case <-ticker.C:
++            fetchSortPersistPeers(bcktr, pstore, host) 
++       }
++
++     lc.Append(fx.Hook{
```

Comment:
Please use fx.Annotate for start and stop instead of hooks

Reply:
Since this was used in the context of an `Invoke` I am not sure what to annotate, since we are dealing with an invocation. Will look into this.

Reply:
example from existing code:

```
	fx.Invoke(fx.Annotate(
				func(
					bcast libhead.Broadcaster[*header.ExtendedHeader],
					fetcher *core.BlockFetcher,
					pubsub *shrexsub.PubSub,
					construct header.ConstructFn,
					store *eds.Store,
				) *core.Listener {
					return core.NewListener(bcast, fetcher, pubsub.Broadcast, construct, store, p2p.BlockTime)
				},
				fx.OnStart(func(ctx context.Context, listener *core.Listener) error {
					return listener.Start(ctx)
				}),
				fx.OnStop(func(ctx context.Context, listener *core.Listener) error {
					return listener.Stop(ctx)
				}),
			)),
```

Reply:
@distractedm1nd I am familiar with that example, except that we'd have to have something to annotate. In this example, I would create a new object type called:
```go
type TimeManager struct {
   ch chan struct{}
   ticker time.Ticker
}
```
And have a function to provide this type, as the thing to be annotated, so that we can convert the `Invoke` style into the `Annotate` style
```go
	fx.Invoke(fx.Annotate(func(ctx context.Context) TimeManager {
  	               return TimeManager{ ticker: time.NewTicker(...) }
                               },
				fx.OnStart(func(ctx context.Context) error {
// launch go routine to periodically call fetchSortPersist...
				}),
				fx.OnStop(func(ctx context.Context) error {
				// trigger routine stop by signaling on channel
				}),
			)),

```